### PR TITLE
Add HMAC authentication to event-api-lambda

### PR DESCRIPTION
**Pull request by kenoir**
## What does this change?

This change adds the ability to authenticate via HMAC using a mechanism compatible with that defined by the [guardian/panda-hmac](https://github.com/guardian/panda-hmac) library. This is intended to allow machine clients to send authenticated telemetry to this service outside of a browser environment.

The Newsroom Resilience team intend to use this mechanism to send metrics from InDesign on the use of various functions so that usage can be better understood, for example [this dashboard](https://metrics.gutools.co.uk/d/krkjrUs4z/octopus-to-composer-events?orgId=1&from=now-7d&to=now). The current mechanism is bypassing the API and uploading telemetry data directly to S3 which is undesirable.

## How to test

You can verify locally that this is behaving as expected by running the API locally using `npm run start` and [using this script](https://gist.github.com/kenoir/1a8ba8f970b36d5f0dd9e646f222153c) to send requests, changing the secret and reviewing API responses.

- [x] Deploy the infrastructure change and updated lambda to the CODE environment.
- [x] Verify requests to the CODE environment proceed as expected with HMAC authentication headers. 

Run `npm run test`:

Screenshot 2023-05-17 at 14 51 51

This change also increases overall test coverage for authentication.

A reviewer should also [review and test the HMAC secret update process](https://github.com/guardian/editorial-tools-user-telemetry-service/blob/6619c783f05a729f51b5cf96872496a58e9aa279/README.md).

## How to deploy?

The appropriate cloudformation stack needs to manually deployed in the `composer` account by running:

```sh
cd cdk
npm run synth
# The CF templates will be in cdk/cdk.out
``` 

The lambdas can then be deployed via Riff-Raff - the CF stack update is non-destructive and can be performed before with event lambda deployment without causing downtime.

## How can we measure success?

InDesign can send metrics to the `editorial-tools-user-telemetry-service` API and these appear on the expected dashboard for ESD to review. We no longer bypass the API and consolidate the methods for sending metric data.

## Have we considered potential risks?

Adding a new mechanism for authentication widens the ways in which this service can be accessed and increases the potential for misuse if the secret key is made public for any reason.

### CODE REVIEW

File: projects/event-lambdas/src/lib/authentication.ts [hunk 1]
```diff
@@ -16,30 +19,70 @@ export const panda = new PanDomainAuthentication(
   guardianValidation
 );
 
+function isHMACValid(date: string, path: string, requestToken: string) {
+  const hmac = createHmac('sha256', hmacSecretKey);
+  const content = date + '\n' + path;
+  hmac.update(content, "utf8");
+
+  return "HMAC " + hmac.digest('base64');
```
Comment (andrew-nowak) [inline projects/event-lambdas/src/lib/authentication.ts:27]:
```suggestion
  return ("HMAC " + hmac.digest('base64')) === requestToken;
```

File: projects/event-lambdas/src/lib/authentication.ts [hunk 2]
```diff
@@ -16,30 +19,70 @@ export const panda = new PanDomainAuthentication(
...
+}
+
+function isDateValid(requestDate: string) {
+  const parsedDate = Date.parse(requestDate);
+  if(parsedDate === NaN) {
+    return false;
+  }
+
+  const currentDate = Date.parse(new Date().toUTCString());
```
Comment (andrew-nowak) [inline projects/event-lambdas/src/lib/authentication.ts:36]:
```suggestion
  const currentDate = new Date().getTime();
```

File: projects/event-lambdas/src/lib/authentication.ts [hunk 3]
```diff
@@ -16,30 +19,70 @@ export const panda = new PanDomainAuthentication(
...
+  const dateDelta = Math.abs(parsedDate - currentDate);
+
+  return dateDelta > hmacAllowedDateOffsetInMillis
```
Comment (andrew-nowak) [inline projects/event-lambdas/src/lib/authentication.ts:39]:
```suggestion
  return dateDelta < hmacAllowedDateOffsetInMillis
```

**Review by jonathonherbert — COMMENTED** [review]
Thanks for the thorough documentation, and for the general spruce of the repo – much appreciated. You've left it in a much better state than you found it!

I'm delighted that we're able to mock secrets-manager in localstack.

Just a few comments, all non-blocking. I haven't yet tested this in CODE, but I thought I'd get those out first – I'll ✅ after I've tested successfully.

**Review by jonathonherbert — APPROVED** [review]
Tested with the script in #25 in CODE with:

- ✅  events signed with a valid key
- ❌  events with an invalid secret
- ❌  events with a valid secret but invalid date

All works as expected.

**Review by jonathonherbert — APPROVED** [review]
Tested with the script in #25 in CODE with:

- ✅  events signed with a valid key
- ❌  events with an invalid secret
- ❌  events with a valid secret but invalid date

All works as expected.

**Review by jonathonherbert — APPROVED** [review]
Tested with the script in #25 in CODE with:

- ✅  events signed with a valid key
- ❌  events with an invalid secret
- ❌  events with a valid secret but invalid date

All works as expected.

**Pull request merged**